### ISSUE TITLE
cmd/go: missing Darcs support

### ISSUE THREAD

**Issue by curiousgay** [golang/go#44471]
Copied from [Wikipedia](https://en.wikipedia.org/wiki/Darcs): Darcs is a distributed version control system created by David Roundy. Key features include the ability to choose which changes to accept from other repositories, interaction with either other local (on-disk) repositories or remote repositories via SSH, HTTP, or email, and an unusually interactive interface. The developers also emphasize the use of advanced software tools for verifying correctness: the expressive type system of the functional programming language Haskell enforces some properties, and randomized testing via QuickCheck verifies many others. The name is a recursive acronym for Darcs Advanced Revision Control System.

**Comment by ianlancetaylor** [comment]
Are there any source code hosting sites that support Darcs?

**Comment by curiousgay** [comment]
@ianlancetaylor currently I only know of https://hub.darcs.net/

**Comment by andrewarchi** [comment]
Related to Darcs is Pijul, with hosting at https://nest.pijul.com/.

**Comment by FrankDMartinez** [comment]
I would greatly appreciate Darcs support as well.

**Comment by LukeMauldin** [comment]
Is there any way to support Golang modules in either Darcs or Pijul without the core Go team adding support? Didn’t know if there were work-around for non-native VCSes? I would think perforce falls into this category too.

**Comment by seankhliao** [comment]
it's possible: the go command understands a "mod" vcs where it will request directly from a proxy. As long as the proxy can get the code and zip it up correctly (and respond to the metadata requests) then it will work

**Comment by LukeMauldin** [comment]
Yes, that makes sense. Thank you for reminding me of that functionality

**Comment by nishanths** [comment]
>  the go command understands a "mod" vcs where it will request directly from a proxy.

For anyone interested you can find more details with `go help goproxy`
and at https://go.dev/ref/mod#goproxy-protocol.

**Comment by FrankDMartinez** [comment]
So, I am at a loss for how this would work. I have a repository [here](https://hub.darcs.net/FrankDMartinez/slice). When I try to `go get -u <this URL>`, I get "unrecognized import path: no go-import meta tags ()". How would I set up my Go configuration to pull this package?

**Comment by nishanths** [comment]
The darcs version control system is not supported by the go command in go1.20.1, the latest release. Quoting from the output of the command `go help importpath`:

> The supported version control systems are:
>
>       Bazaar      .bzr
>       Fossil      .fossil
>       Git         .git
>       Mercurial   .hg
>       Subversion  .svn

Once darcs is a supported version control system in the go command, additionally one of the following must happen before you can use `go get` with the repository: 

* the go command must to be taught to handle the hub.darcs.net code hosting service [*]; or 
* the  hub.darcs.net repository URL must serve the correct `<meta name="go-import">` tags in order to be able to fetch the code. \[**] 

\[*] Currently the go command only knows of Bitbucket, GitHub, Launchpad, and IBM DevOps Services. 

\[**] See the output of `go help importpath`.

**Comment by FrankDMartinez** [comment]
So, if I understand correctly, I have to wait until either Darcs Hub adds support for the meta tag or Go adds support for Darcs Hub without requiring the tag?

**Comment by nishanths** [comment]
Correct. 

There is another way. You can still use a darcs repository and Darcs Hub, but instead of an import path starting with "hub.darcs.net", you could use a custom import path with a domain that you control. For example:

```
go get code.mydomain.com/slice
```

Then at the URL `https://code.mydomain.com/slice?go-get=1` you can serve either (or both?) the vcs variant and the mod variant of the `<meta>` tags. For the vcs variant of the `<meta>` tag to work successfully, the go command would still need to support darcs as a version control system, so this might involve waiting. 

On the other hand, for the mod variant of the `<meta>` tag, you could set up your own Go module proxy server (say, at `https://goproxy.mydomain.com`) that serves the module zip file representation of the code in the Darcs Hub repository.
